### PR TITLE
[Refactor] Refactor rewriter context for MMRazor

### PR DESCRIPTION
**Pull request by grimoire**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Refactor rewriter for https://github.com/open-mmlab/mmdeploy/pull/1471
`ctx` has been removed from arguments.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by grimoire** [comment]
@pppppM please take a look.

File: mmdeploy/apis/onnx/export.py [hunk 1]
```diff
@@ -116,6 +116,15 @@ def _add_or_update(cfg: dict, key: str, val: Any):
                 input_metas, dict
             ), f'Expect input_metas type is dict, get {type(input_metas)}.'
             model_forward = patched_model.forward
+
```
Comment (RunningLeon) [inline mmdeploy/apis/onnx/export.py:119]: @grimoire some ci failed, will it be fixed in this PR?

File: mmdeploy/mmcv/cnn/conv2d_adaptive_padding.py [hunk 2]
```diff
@@ -1,86 +0,0 @@
...
-            'Constant', value_t=torch.tensor(0, dtype=torch.int64))
-        return g.op(
-            'Pad', x, padded, constant_value, mode_s='constant', outputs=1)
-
-
-@FUNCTION_REWRITER.register_rewriter(
-    func_name='mmcv.cnn.bricks.conv2d_adaptive_padding. \
-        Conv2dAdaptivePadding.forward',
-    backend=Backend.TENSORRT.value)
-def conv2d_adaptive_padding__forward__tensorrt(ctx, self, x):
```
Comment (RunningLeon) [inline mmdeploy/mmcv/cnn/conv2d_adaptive_padding.py:60]: why do we remove this rewriting?

File: mmdeploy/apis/onnx/optimizer.py [hunk 3]
```diff
@@ -5,8 +5,9 @@
 
 
 @FUNCTION_REWRITER.register_rewriter('torch.onnx.utils._model_to_graph')
-def model_to_graph__custom_optimizer(ctx, *args, **kwargs):
+def model_to_graph__custom_optimizer(*args, **kwargs):
     """Rewriter of _model_to_graph, add custom passes."""
+    ctx = FUNCTION_REWRITER.get_context()
```
Comment (RunningLeon) [inline mmdeploy/apis/onnx/optimizer.py:10]: could update the `rewriting demo code` in docs as well,
for instance , this doc:
https://github.com/open-mmlab/mmdeploy/edit/dev-1.x/docs/en/07-developer-guide/support_new_model.md

File: mmdeploy/codebase/mmdet/models/dense_heads/base_dense_head.py [hunk 4]
```diff
@@ -237,6 +236,8 @@ def base_dense_head__predict_by_feat__rknn(
             tuple[Tensor, Tensor, Tensor]: batch_mlvl_bboxes,
                 batch_mlvl_scores, batch_mlvl_centerness
     """
+    ctx = FUNCTION_REWRITER.get_context()
```
Comment (RunningLeon) [inline mmdeploy/codebase/mmdet/models/dense_heads/base_dense_head.py:239]: ssd+tensorrt conversion failed
```
  File "/root/workspace/mmdeploy/mmdeploy/apis/onnx/export.py", line 123, in wrapper
    return forward(*arg, **kwargs)
  File "/root/workspace/mmdeploy/mmdeploy/codebase/mmdet/models/detectors/single_stage.py", line 74, in single_stage_detector__forward
    return __forward_impl(
  File "/root/workspace/mmdeploy/mmdeploy/core/optimizers/function_marker.py", line 266, in g
    rets = f(*args, **kwargs)
  File "/root/workspace/mmdeploy/mmdeploy/codebase/mmdet/models/detectors/single_stage.py", line 24, in __forward_impl
    output = self.bbox_head.predict(x, data_samples, rescale=False)
  File "/opt/conda/envs/torch1.10.0/lib/python3.8/site-packages/mmdet/models/dense_heads/base_dense_head.py", line 197, in predict
    predictions = self.predict_by_feat(
  File "/root/workspace/mmdeploy/mmdeploy/codebase/mmdet/models/dense_heads/base_dense_head.py", line 121, in base_dense_head__predict_by_feat
    priors = pad_with_value_if_necessary(priors, 1, pre_topk)
  File "/root/workspace/mmdeploy/mmdeploy/codebase/mmdet/deploy/utils.py", line 142, in pad_with_value_if_necessary
    return __pad_with_value_if_necessary(
TypeError: __pad_with_value_if_necessary__tensorrt() missing 1 required positional argument: 'pad_dim'
12/05 03:54:13 - mmengine - ERROR - /root/workspace/mmdeploy/mmdeploy/apis/core/pipeline_manager.py - pop_mp_output - 80 - `mmdeploy.apis.pytorch2onnx.torch2onnx` with Call id: 0 failed. exit.
```

**Comment by RunningLeon** [comment]
This PR would affect the deployment of mmyolo, may let @hhaAndroid  be aware of this.

File: mmdeploy/mmcv/cnn/conv2d_adaptive_padding.py (hunk 2 above)
Comment (grimoire) [reply mmdeploy/mmcv/cnn/conv2d_adaptive_padding.py:60]: `'mmcv.cnn.bricks.conv2d_adaptive_padding. \
        Conv2dAdaptivePadding.forward'` will not map to any function. This rewriter has never been enabled.

Comment (RunningLeon) [reply mmdeploy/mmcv/cnn/conv2d_adaptive_padding.py:60]: It is added by @SsTtOoNnEe for deployment of mmcls::efficientnet for torch>=1.10
https://github.com/open-mmlab/mmcv/blob/2478cb6f840206eb0d78cd27320d9068be483860/mmcv/cnn/bricks/conv2d_adaptive_padding.py#L46

Comment (grimoire) [reply mmdeploy/mmcv/cnn/conv2d_adaptive_padding.py:60]: The rewriter has never been enabled. The func_name string is WRONG.

Comment (RunningLeon) [reply mmdeploy/mmcv/cnn/conv2d_adaptive_padding.py:60]: then we could fix the `func_name` instead of removing the rewriting.

Comment (grimoire) [reply mmdeploy/mmcv/cnn/conv2d_adaptive_padding.py:60]: Enable the rewriter would failed the export of efficientnet

**Review by RunningLeon — APPROVED** [review]
LGTM

**Review by hanrui1sensetime — APPROVED** [review]
LGTM.

File: mmdeploy/core/rewriters/rewriter_utils.py [hunk 5]
```diff
@@ -379,3 +382,61 @@ def wrapper(*args, **kwargs):
...
+    g_vars = frame.f_globals
+    func_name = frameinfo.function
+    assert func_name in g_vars, \
+        f'Can not find function: {func_name} in global.'
+    func = g_vars[func_name]
+    return func
+
+
+def get_frame_qual_name(top: int = 1) -> str:
+    """get frame name."""
```
Comment (lvhan028) [inline mmdeploy/core/rewriters/rewriter_utils.py:414]: vote for `get_frame_qualname` as well

**Comment by lvhan028** [comment]
`1.x` is mmdeploy's release branch while `dev-1.x` is its developing branch.
So please change the target branch to `open-mmlab:dev-1.x`

**Comment by lvhan028** [comment]
@pppppM any comments?

File: mmdeploy/core/rewriters/function_rewriter.py [hunk 6]
```diff
@@ -1,8 +1,11 @@
 # Copyright (c) OpenMMLab. All rights reserved.
+import types
 from collections import defaultdict
 from typing import (Any, Callable, Dict, List, MutableSequence, Optional,
                     Tuple, Union)
 
+from torch.fx._symbolic_trace import _wrapped_fns_to_patch
```
Comment (grimoire) [inline mmdeploy/core/rewriters/function_rewriter.py:7]: In torch<=1.9, `torch.fx._symbolic_trace` does not exist and  `_wrapped_fns_to_patch` is inside `torch.fx.symbolic_trace`.

Comment (grimoire) [reply mmdeploy/core/rewriters/function_rewriter.py:7]: Nope, update `_wrapped_fns_to_patch` is not necessary if we do not need to trace fx. And we can import `_wrapped_fns_to_patch` from `torch.fx.symbolic_trace` in old torch version.

Comment (lvhan028) [reply mmdeploy/core/rewriters/function_rewriter.py:7]: But UT failed
https://app.circleci.com/pipelines/github/open-mmlab/mmdeploy/3442/workflows/1a4e1734-ab69-43c9-a7f6-edb8e8a7cdc2/jobs/9419

Comment (grimoire) [reply mmdeploy/core/rewriters/function_rewriter.py]: [This](https://github.com/open-mmlab/mmdeploy/pull/1483/commits/392946a632f83db28008ecb00f6d5d7e8dd3f9d2) change is commited by @pppppM , comment here to discuss the detail.

**Pull request merged**